calyx-opt/src/passes/static_inliner.rs

```diff
@@ -0,0 +1,388 @@
+                // body group. So the static group will literally look like:
+                // static group static_repeat <num_repeats * body_latency> {body[go] = 1'd1;}
+                structure!( builder;
+                    let signal_on = constant(1,1);
+                );
+                let trigger_body = build_assignments!(builder;
+                    body_group["go"] = ? signal_on["out"];
+                );
+                repeat_group.borrow_mut().assignments = trigger_body.to_vec();
+                repeat_group.borrow_mut().attributes = attributes.clone();
+                repeat_group
+            }
+            ir::StaticControl::Empty(_) => unreachable!(
+                "should not call inline_static_control on empty stmt"
+            ),
+            ir::StaticControl::Invoke(_) => {
+                todo!("implement static inlining for invokes")
+            }
+        }
+    }
+
+    // searches thru `con` for "static islands"
+    // when it finds a "static island", then creates a corresponding
+    // static group by calling `inline_static_control`
+    // thne adds entry (id of "static island" control, equivalent static group)
+    // to self.map
+    fn build_static_map(
+        &mut self,
+        con: &ir::Control,
+        builder: &mut ir::Builder,
+    ) {
+        match con {
+            ir::Control::Enable(_)
+            | ir::Control::Empty(_)
+            | ir::Control::Invoke(_) => (),
+            ir::Control::Seq(ir::Seq { stmts, .. })
+            | ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    self.build_static_map(stmt, builder);
+                }
+            }
+            ir::Control::If(ir::If {
+                tbranch, fbranch, ..
+            }) => {
+                self.build_static_map(tbranch, builder);
+                self.build_static_map(fbranch, builder);
+            }
+            ir::Control::While(ir::While { body, .. }) => {
+                self.build_static_map(body, builder);
+            }
+            ir::Control::Static(sc) => {
+                let id = ControlId::get_guaranteed_id_static(sc);
+                let sgroup = Self::inline_static_control(sc, builder);
+                self.map.insert(id, sgroup);
+            }
+        }
+    }
+}
+
+impl Visitor for StaticInliner {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        sigs: &LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        // assign unique ids so we can use them in our map
+        ControlId::compute_unique_ids(&mut comp.control.borrow_mut(), 0, false);
+        let control_ref = Rc::clone(&comp.control);
+        let mut builder = ir::Builder::new(comp, sigs);
+        // builds static map, which maps static islands to equivalent singular inlined static groups
+        self.build_static_map(&control_ref.borrow(), &mut builder);
```

Comment:
The way this pass is structured, it traverses the control program twice. Once during `build_static_map` and another time in the visitor itself. Can we combine the two? Or rather, is there a reason that `build_static_map` needs to be wholly computed before traversing the program?